### PR TITLE
Remove lodash merge

### PR DESCRIPTION
**Pull request by AndrewMusgrave**
### WHY are these changes introduced?

Part of removing lodash #991 

### WHAT is this pull request doing?

Creating a less featureful `merge` function and using that instead of `_.merge`

### How to 🎩
Test pass? 💯 

### TODO:
* add changelog

### CODE REVIEW

**Comment by BPScott** [comment]
Repeating my query from https://github.com/Shopify/polaris-react/pull/1010 here too:

I'm not too familiar with how mergeAppProviderOptions() and legacy context stuff does its thing, but do we need a deep merge there  and in the withAppProvider / withSticky HoCs or could we get away with a shallow one using Object.assign()?

**Comment by AndrewMusgrave** [comment]
Unfortunately merging context requires a deep merge since the user could define partial context and have it collide with `polaris`.

File: src/utilities/merge.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  return final;
+}
+
+function mergeRecursively(objA: GeneralObject, objB: GeneralObject) {
+  for (const key in objB) {
+    if (!objB.hasOwnProperty(key)) {
+      continue;
+    } else if (isMergableValue(objB[key]) && isMergableValue(objA[key])) {
```
Comment (solonaarmstrong) [inline src/utilities/merge.ts]: `isMergeableValue` ? Sorry to pick on spelling, I just find it easier to understand.

File: src/utilities/merge.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  for (const obj of objs) {
+    mergeRecursively(final, obj);
+  }
+
+  return final;
+}
+
+function mergeRecursively(objA: GeneralObject, objB: GeneralObject) {
+  for (const key in objB) {
+    if (!objB.hasOwnProperty(key)) {
```
Comment (solonaarmstrong) [inline src/utilities/merge.ts]: Would it be worth testing this?

```
   it('returns the first argument if it receives only one argument', () => {
    const objA = {keyA: 1};
    expect(merge(objA)).toEqual(objA);
  });
```

**Comment by solonaarmstrong** [comment]
Elegant! 👌

**Comment by AndrewMusgrave** [comment]
Ready for 👀

**Comment by dleroux** [comment]
I rebased it because conflicts with Michelle's autobind.

**Comment by dleroux** [comment]
I assume we skip change logs for these?

**Comment by dleroux** [comment]
Although I wonder why percy changed.

**Comment by AndrewMusgrave** [comment]
> I assume we skip change logs for these?

Iv'e been adding them to `Code Quality` 😄

**Pull request merged**